Move the default profiles to the user’s home

thufschmitt opened the pull request:
Rather than using `/nix/var/nix/{profiles,gcroots}/per-user/`, put the user
profiles and gcroots under `$XDG_DATA_DIR/nix/{profiles,gcroots}`.

This means that the daemon no longer needs to manage these paths itself
(they are fully handled client-side). In particular, it doesn’t have to
`chown` them anymore (removing one need for root).

This does change the layout of the gc-roots created by nix-env, and is
likely to break some stuff, so I’m not sure how to properly handle that.

First step towards #5208

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-17/15037/1

fricklerhandwerk commented:
Discussed in Nix team meeting on 2022-12-19.

Decision: take out the migration into a different PR, and review this one together when it's ready.

<details><summary>Complete discussion</summary>

- @thufschmitt: changing the directory layout may be breaking for existing tools, but not sure who is relying on this
    - @edolstra: migrating the profile dir precludes rolling back
        - @Ericson2314: we could prefer the old dir if it exists
            - @thufschmitt: then we're stuck with these two locations forever (which we already are)
            - @Ericson2314: at least for a few migration cycles
            - @thufschmitt: will also break third-party tools depending on when Nix first was installed, even if for the same version of Nix. this is the same problem as with NixOS `config.stateVersion`.
                - @infinisil: what about introducing a tool to query for profile paths so tools don't have to rely on actual file system paths?
    - @infinisil: can put a symlink
        - deamon has to be root to follow it
- @edolstra: the daemon should not trust user directories
    - @thufschmitt: this PR does not change anything about how GC works, there is nothing there which cannot already be done by e.g. setting `--profile`
- agreement: take out the migration into a different PR, and review it together next time

</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2022-12-16-nix-team-meeting-minutes-17/24120/1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2022-12-19-nix-team-meeting-minutes-18/24121/1

thufschmitt commented:
I've removed the migration code, now any existing installation will keep using the old profiles directory.

I've also changed the behaviour for `root` to keep the `/nix/var/nix/profiles/default` location as it's also used as a system-wide profile (so it should stay world-readable and stable)

edolstra commented on src/libstore/build/local-derivation-goal.cc:
```diff
@@ -1517,7 +1517,7 @@ void LocalDerivationGoal::startDaemon()
                     daemon::processConnection(store, from, to,
                         daemon::NotTrusted, daemon::Recursive,
-                        [&](Store & store) { store.createUser("nobody", 65535); });
+                        [&](Store & store) {});
```
I think we can remove the `authHook` argument to `processConnection()` now, since it's only used to call `createUser()`.

edolstra commented:
This is probably important enough to mention in the release notes, even if it shouldn't break anything.

domenkozar commented:
See https://github.com/cachix/cachix-action/issues/137

ncfavier commented:
This makes the assumptions in https://github.com/nix-community/home-manager/blob/master/modules/lib-bash/activation-init.sh invalid, causing failures like https://github.com/nix-community/home-manager/actions/runs/4301868239/jobs/7499628919#step:8:260. How do we migrate smoothly?

fricklerhandwerk commented:
@ncfavier please open an issue and ping me, we'll discuss it on the team.

rycee commented:
I'm looking at how to adapt Home Manager to this change. Should I use some HM specific directory to keep the profile and gcroot? E.g. `$XDG_DATA_DIR/home-manager/{profiles,gcroots}`. And nix-collect-garbage will automatically handle the profile but leave the gcroot alone? Or should I put the HM profile and gcroot in `$XDG_DATA_DIR/nix/{profiles,gcroots}`?

Also, this commit doesn't seem to contain any documentation updates. Are those in a separate PR?

rycee commented:
Hmm, I had a look at `nix-collect-garbage.cc` and it seems to me that it will only delete old profile generations if they are present in the Nix state dir? If so, won't it also fail to expire old user profiles after this migration? I'm not very familiar with the Nix code, though, so I may be completely mistaken here.

thiagokokada commented:
> This makes the assumptions in https://github.com/nix-community/home-manager/blob/master/modules/lib-bash/activation-init.sh invalid, causing failures like https://github.com/nix-community/home-manager/actions/runs/4301868239/jobs/7499628919#step:8:260. How do we migrate smoothly?

Interesting. I just hit this issue when trying to use the Nix static inside a Steam Deck, with the store inside `$HOME` (introduced in https://discourse.nixos.org/t/nix-2-10-0-released/20291). While it is unfortunate that this bug, having this being fixed can be exciting.

fricklerhandwerk commented:
@Rycee Apparently this still needs more work on our side. Please open an issue so we can discuss it on the team.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nix-team-report-2022-10-2023-03/27486/1
